nucleus/payara-modules/opentracing-adapter/src/main/java/fish/payara/opentracing/OpenTracingService.java

```diff
@@ -123,28 +129,39 @@ public synchronized Tracer getTracer(String applicationName) {
 
         // If there isn't a tracer for the application, create one
         if (tracer == null) {
-            // Check which type of Tracer to create
+            tracer = createTracer(applicationName);
+        }
 
-            try {//See if an alternate implementation of Tracer is available in a library.
-                ServiceLoader<Tracer> tracerLoader = ServiceLoader.load(Tracer.class);
-                Iterator<Tracer> loadedTracer = tracerLoader.iterator();
-                if (loadedTracer.hasNext()) {
-                    tracer = loadedTracer.next();
-                }
-            } catch (NoClassDefFoundError ex) {
-                logger.log(Level.SEVERE, "Unable to find Tracer implementation", ex);
-            }
+        return tracer;
+    }
 
-            if (Boolean.getBoolean("USE_OPENTRACING_MOCK_TRACER")) {
-                tracer = new MockTracer(new ThreadLocalScopeManager(), MockTracer.Propagator.TEXT_MAP);
-            } else if (tracer == null) {
-                tracer = new fish.payara.opentracing.tracer.Tracer(applicationName, scopeManager);
+    private synchronized Tracer createTracer(String applicationName) {
+        // Does this NEED to be synchronised? Tracers don't store state, and Scopes are ThreadLocal
+
+        // Double-checked locking - potentially naughty
+        Tracer tracer = tracers.get(applicationName);
```

Comment:
How about ConcurrentHashMap.computeIfAbsent and put the following code into the compute function.